[WIP] Refactor gce module to be more flexible

detiber opened the pull request:
- [ ] Allow for creating multiple instances without specifying a list of instance names
- [x] Allow more flexible disk definitions for instances, including local-ssd scratch disks
- [ ] Update termination code to support using the ex_destroy_multiple_nodes method
- [ ] Support deleting multiple instances by tag
- [ ] Support deleting one or more instances by instance ids (or some other identifier other than name)

detiber committed:
Start of refactor to support more flexible disk definition and multiple instance creation without a list of instance names

erjohnso commented on cloud/google/gce.py:
```diff
@@ -423,7 +378,91 @@ def terminate_instances(module, gce, instance_names, zone_name):
+    return (changed, instance_json_data, [ i['name'] for i in instance_json_data ])
+
+def create_instances_by_count(module, gce):
+    # TODO: implement using ex_create_multiple_nodes
```
That underlying libcloud function may need some attention.  I don't think I've been doing a stellar job at keeping it updated.

erjohnso commented on cloud/google/gce.py:
```diff
@@ -423,7 +378,91 @@ def terminate_instances(module, gce, instance_names, zone_name):
+def cleanup_metadata(metadata):
+    if metadata:
+        try:
+            md = literal_eval(metadata)
```
Please see #154.  Should be,

```
md = literal_eval(str(metadata))
```

detiber committed:
Move functionality into a class, some more groundwork for changes, still
non-functioning.

detiber commented:
@erjohnso Many thanks for the initial review...  Latest updates are still non-functioning but since I've started to encapsulate the functionality into a class I wanted to get your thoughts on the slightly new direction I was taking it.

detiber committed:
Actually implement something

- implemented create_instance_by_name

- updated params
  - remove persistent_boot_disk flag, since when set to false, the libcloud
    driver still creates a persistent disk, and when set to true, the behavior
    is the same as the libvirt default when ex_disks_gce_struct and
    ex_boot_disk are both not set
  - added boot_disk, disk_type, use_existing_disk and nics params

detiber committed:
implement terminate and fix issue with instance_names param when string

detiber committed:
workaround for libcloud create_node not properly taking a str for ex_boot_disk

detiber committed:
updates

- don't bite off so much for new features (stopping instances, exact_count for
  instances, etc)
- clean up disks attribute to not force user to provide full path for
  attributes
  - still some issues that are preventing the ex_gce_disks_struct from working
    properly...

detiber committed:
workaround for ex_get_disktype not seeming to work properly and updates for documentation and examples

detiber committed:
backwards compat for disks param

detiber committed:
removing nics parameter, since the libvirt functionality is currently broken

flyinprogrammer commented:
:+1: to these changes - how can I help?

detiber commented:
@flyinprogrammer sorry for the delay, been heads down on a separate project this past month and haven't had much time to pick this back up.  I was hitting some snags with libcloud bugs the last time I was poking around though.  Feel free to pick this up and run with it if you have the time, as I suspect it will be a bit before I get a chance to dig back into this.

gregdek commented:
OK, since @detiber has stated that he doesn't have time for this one, closing this PR -- @flyinprogrammer recommend you pick this up and submit as a new PR if you're interested in getting it in.
